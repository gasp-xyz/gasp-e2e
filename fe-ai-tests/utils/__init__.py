# This file makes the utils directory a Python package
# This allows imports like: from utils.common_test_utils import x
