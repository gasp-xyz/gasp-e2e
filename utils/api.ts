import { ApiPromise } from "@polkadot/api";
import { testLog } from "./Logger";
import { getEnvironmentRequiredVars, getMangataApiUrlPort } from "./utils";
import { Mangata, MangataInstance } from "@mangata-finance/sdk";
import { getPort } from "get-port-please";
import XcmNetworks from "./Framework/XcmNetworks";
import { BuildBlockMode } from "@acala-network/chopsticks";
import { ApiContext } from "./Framework/XcmHelper";

export let api: ApiPromise | null = null;
export let mangata: MangataInstance | null = null;
export let mangataChopstick: ApiContext | null = null;
export let chopstickUri: String;
export const getApi = () => {
  if (!api) {
    throw new Error("Api not initialized");
  }
  return api;
};

export const initApi = async (uri = "") => {
  if (!uri) {
    const { chainUri: envUri } = getEnvironmentRequiredVars();
    uri = envUri;
  }
  if (process.env.CHOPSTICK_ENABLED) {
    const mgaPort = getMangataApiUrlPort();
    const chopstickPort = await getPort();
    mangataChopstick = await XcmNetworks.mangata({
      localPort: chopstickPort,
      buildBlockMode: BuildBlockMode.Instant,
    });
    uri = uri.replace(mgaPort.toString(), chopstickPort.toString());
    chopstickUri = uri;
  }
  testLog.getLog().info(`TEST_INFO: Running test in ${uri}`);
  mangata = Mangata.instance([uri]);
  api = await mangata.api();
  return api;
};

export async function getMangataInstance(uri = ""): Promise<MangataInstance> {
  if (!api) {
    await initApi(uri);
  }
  return mangata!;
}
export const disconnect = async () => {
  //TODO:sdkV2
  //  if (mangata) {
  //    await mangata.disconnect();
  //  } else {
  //    await Mangata.instance([uri]).disconnect();
  //  }
};
