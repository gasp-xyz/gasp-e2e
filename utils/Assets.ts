import { BN_TEN, BN_THOUSAND } from "@mangata-finance/sdk";
import { ApiPromise } from "@polkadot/api";
import { AccountInfo } from "@polkadot/types/interfaces";
import { BN } from "@polkadot/util";
import { assert } from "console";
import _ from "lodash";
import { MGA_ASSET_ID } from "./Constants";
import { ExtrinsicResult } from "./eventListeners";
import { api, Extrinsic, setupApi, setupUsers } from "./setup";
import { Sudo } from "./sudo";
import { getAssetSupply, getNextAssetId } from "./tx";
import {
  getEventResultFromMangataTx,
  setAssetInfo,
  sudoIssueAsset,
} from "./txHandler";
import { User } from "./User";
import { MangataTypesAssetsCustomMetadata } from "@polkadot/types/lookup";

export class Assets {
  static MG_UNIT: BN = BN_TEN.pow(new BN(18));
  static DEFAULT_AMOUNT = BN_THOUSAND.mul(this.MG_UNIT);

  ///This method create or return the specified number of available assets
  static async getCurrencies(numAssets: number = 2, sudoUser: User) {
    const currencies: string[] = [];
    const numberOfcurrencies = parseInt((await getNextAssetId()).toString());

    if (numAssets > numberOfcurrencies) {
      //we need to create some currencies.
      for (
        let remainingAssetsToCreate = numberOfcurrencies;
        remainingAssetsToCreate < numAssets;
        remainingAssetsToCreate++
      ) {
        await this.issueAssetToSudo(sudoUser);
      }
    }
    //there are some currencies already created.
    for (let index = 0; index < numAssets; index++) {
      await getAssetSupply(new BN(index));
      currencies.push(index.toString());
    }

    return currencies;
  }

  static async setupUserWithCurrencies(
    user: User,
    currencyValues = [new BN(250000), new BN(250001)],
    _sudo: User,
    skipInfo = false
  ): Promise<BN[]> {
    const txs: Extrinsic[] = [];
    await setupApi();
    await setupUsers();
    for (let currency = 0; currency < currencyValues.length; currency++) {
      txs.push(Assets.issueToken(user, currencyValues[currency]));
    }
    const result = await Sudo.batchAsSudoFinalized(...txs);
    const assetIds: BN[] = result
      .filter((X) => X.method === "Issued")
      .map((t) => new BN(t.eventData[0].data.toString()));
    const addInfos: Extrinsic[] = [];
    if (!skipInfo) {
      for (let index = 0; index < assetIds.length; index++) {
        const assetId = assetIds[index];
        addInfos.push(
          Assets.registerAsset(
            `TEST_${assetId}`,
            this.getAssetName(assetId.toString()),
            new BN(18)
          )
        );
      }
      await Sudo.batchAsSudoFinalized(...addInfos);
    }
    user.addAssets(assetIds);
    await user.refreshAmounts();
    return assetIds;
  }

  static async issueAssetToSudo(sudo: User) {
    await this.issueAssetToUser(sudo, new BN(1000), sudo);
  }

  //this method add a certain amount of currencies to a user into a returned currecncyId
  static async issueAssetToUser(
    user: User,
    num = new BN(1000),
    sudo: User,
    skipInfo = false
  ) {
    const result = await sudoIssueAsset(
      sudo.keyRingPair,
      num,
      user.keyRingPair.address
    );
    const eventResult = await getEventResultFromMangataTx(result, [
      "tokens",
      "Issued",
      user.keyRingPair.address,
    ]);

    assert(eventResult.state === ExtrinsicResult.ExtrinsicSuccess);
    const assetId = eventResult.data[0].split(",").join("");
    if (!skipInfo) {
      await setAssetInfo(
        sudo,
        new BN(assetId),
        `TEST_${assetId}`,
        this.getAssetName(assetId),
        `Test token ${assetId}`,
        new BN(18)
      );
    }

    return new BN(assetId);
  }

  static getAssetName(assetID: string) {
    return `m${assetID}`;
  }

  static mintNative(user: User, amount: BN = this.DEFAULT_AMOUNT): Extrinsic {
    user.addAsset(MGA_ASSET_ID);
    return Sudo.sudo(
      api.tx.tokens.mint(MGA_ASSET_ID, user.keyRingPair.address, amount)
    );
  }

  static issueToken(user: User, amount: BN = this.DEFAULT_AMOUNT): Extrinsic {
    return Sudo.sudo(api.tx.tokens.create(user.keyRingPair.address, amount));
  }

  static mintToken(
    asset: BN,
    user: User,
    amount: BN = this.DEFAULT_AMOUNT
  ): Extrinsic {
    return Sudo.sudo(
      api.tx.tokens.mint(asset, user.keyRingPair.address, amount)
    );
  }

  static transfer(target: User, tokenId: BN, amount: BN): Extrinsic {
    return api.tx.tokens.transfer(target.keyRingPair.address, tokenId, amount);
  }

  static transferAll(target: User, tokenId: BN): Extrinsic {
    return api.tx.tokens.transferAll(target.keyRingPair.address, tokenId, true);
  }
  static FinalizeTge(): Extrinsic {
    return Sudo.sudo(api!.tx.issuance.finalizeTge());
  }
  static initIssuance(): Extrinsic {
    return Sudo.sudo(api!.tx.issuance.initIssuanceConfig());
  }
  static promotePool(liquidityId: number, weight: number = 20): Extrinsic {
    return Sudo.sudo(
      api!.tx.proofOfStake.updatePoolPromotion(liquidityId, weight)
    );
  }
  static registerAsset(
    name: string,
    symbol: string,
    decimals: number | BN,
    // location?: MultiLocation,
    location?: object,
    xcmMetadata?: XcmMetadata,
    xykMetadata?: XykMetadata,
    assetId?: number | BN
  ): Extrinsic {
    return Sudo.sudo(
      api.tx.assetRegistry.registerAsset(
        {
          decimals: decimals,
          name: api.createType("Vec<u8>", name),
          symbol: api.createType("Vec<u8>", symbol),
          existentialDeposit: 0,
          location: location ? { V1: location } : null,
          additional: {
            xcm: xcmMetadata,
            xyk: xykMetadata,
          },
        },
        //@ts-ignore
        api.createType("Option<u32>", assetId)
      )
    );
  }

  static updateAsset(assetId: number | BN, update: UpdateAsset): Extrinsic {
    const optional = update.decimals
      ? api.createType("Option<u32>", update.decimals)
      : null;
    return Sudo.sudo(
      api.tx.assetRegistry.updateAsset(
        assetId,
        //@ts-ignore
        optional,
        api.createType("Vec<u8>", update.name),
        api.createType("Vec<u8>", update.symbol),
        null,
        update.location
          ? update.location.location
            ? { V2: update.location } // Some(location)
            : api.createType("Vec<u8>", "0x0100") // Some(None)
          : null, // None
        update.metadata!
      )
    );
  }
  static async disableToken(tokenId: BN) {
    const isRegistered = (
      await api!.query.assetRegistry.metadata(tokenId)
    ).toHuman();
    let extrinsicToDisable: Extrinsic;
    if (!isRegistered) {
      extrinsicToDisable = Assets.registerAsset(
        "testAsset" + tokenId.toString(),
        "testSymbol",
        18,
        undefined,
        undefined,
        {
          operationsDisabled: true,
        },
        tokenId
      );
    } else {
      const metadata: MangataTypesAssetsCustomMetadata = api.createType(
        "MangataTypesAssetsCustomMetadata",
        {
          xyk: {
            operationsDisabled: true,
          },
        }
      );
      extrinsicToDisable = Assets.updateAsset(tokenId, {
        metadata: metadata,
      });
    }
    return await Sudo.asSudoFinalized(extrinsicToDisable);
  }
  static async enableToken(tokenId: BN) {
    const isRegistered = (
      await api!.query.assetRegistry.metadata(tokenId)
    ).toHuman();
    let extrinsicToTokenEnable: Extrinsic;
    if (
      isRegistered &&
      _.get(isRegistered, "additional.xyk.operationsDisabled")
    ) {
      const metadata: MangataTypesAssetsCustomMetadata = api.createType(
        "MangataTypesAssetsCustomMetadata",
        {
          xyk: {
            operationsDisabled: false,
          },
        }
      );
      extrinsicToTokenEnable = await Assets.updateAsset(tokenId, {
        metadata: metadata,
      });
      await Sudo.asSudoFinalized(extrinsicToTokenEnable);
    }
  }
}

export const balance = async (api: ApiPromise, address: string) => {
  const account = await api.query.system.account<AccountInfo>(address);
  return account.data.toJSON();
};

interface XcmMetadata {
  feePerSecond: number;
}

interface XykMetadata {
  operationsDisabled: boolean;
}

// API expects Option<Option<Location>>, we need to wrap it so we can pass
// undefined in UpdateAsset - no update
// { location: undefined } - update with no location
// { location: { ... } } - update with some location
interface UpdateLocation {
  location?: object;
}

// if the value is undefined, it will not be updated
interface UpdateAsset {
  name?: string;
  symbol?: string;
  decimals?: number;
  // location?: MultiLocation,
  location?: UpdateLocation;
  metadata?: MangataTypesAssetsCustomMetadata;
}
