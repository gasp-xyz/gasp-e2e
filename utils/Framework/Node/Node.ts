import * as uuid from "uuid";
import { ApiPromise } from "@polkadot/api";
import { initApi } from "../../api";
import { testLog } from "../../Logger";
import { BN } from "@polkadot/util";
import { GovernanceUser } from "../User/GovernanceUser";

export class Node {
  name: string;
  wsPath: string;
  api?: ApiPromise;

  firstBlock?: number;
  lastBlock?: number;
  lastHash?: string;
  hashes: Set<string> = new Set();
  blockHashes: Map<number, string> = new Map();
  subscription: any;

  electionEvents: Map<number, { candidates: any; members: any }> = new Map();
  userBalancesHistory: Map<
    number,
    Map<number, { free: BN; reserved: BN; miscFrozen: BN; feeFrozen: BN }>
  > = new Map();
  systemExtrinics: any[] = [];
  constructor(wsPath: string) {
    this.name = uuid.v4();
    this.wsPath = wsPath;
  }

  async connect(): Promise<void> {
    this.api = await initApi(this.wsPath);
  }

  async subscribeToHead(): Promise<void> {
    this.subscription = await this.api!.rpc.chain.subscribeNewHeads(
      async (lastHeader) => {
        if (!this.firstBlock) {
          this.firstBlock = lastHeader.number.toNumber();
        }
        this.lastBlock = lastHeader.number.toNumber();
        this.lastHash = lastHeader.hash.toString();
        this.hashes.add(this.lastHash);
        this.blockHashes.set(this.lastBlock, this.lastHash);
        testLog
          .getLog()
          .debug(`${this.name} - #${this.lastBlock} - ${this.lastHash}`);
        await this.api!.queryMulti(
          [
            this.api!.query.elections.candidates,
            this.api!.query.elections.members,
          ],
          ([candidates, members]) => {
            this.electionEvents.set(this.lastBlock!, {
              candidates: candidates.toJSON(),
              members: members.toJSON(),
            });

            testLog.getLog().debug(
              `Saved Election Information 
               Candidates: ${this.electionEvents.get(this.lastBlock!)
                 ?.candidates}
               Members: ${this.electionEvents.get(this.lastBlock!)?.members}`,
            );
          },
        );
      },
    );
  }
  async subscribeToUserBalanceChanges(
    candidate: GovernanceUser,
  ): Promise<void> {
    this.subscription = await this.api!.rpc.chain.subscribeNewHeads(
      async (lastHeader) => {
        const balancesAtblock = await candidate.getAllUserTokens();
        this.userBalancesHistory.set(
          lastHeader.number.toNumber(),
          balancesAtblock,
        );
      },
    );
  }
  async subscribeToExtrinsics(): Promise<void> {
    this.subscription = await this.api!.rpc.chain.subscribeNewHeads(
      async (lastHeader) => {
        const blockNo = lastHeader.number.toBn();
        const blockHash = await this.api!.rpc.chain.getBlockHash(blockNo);
        const signedBlock = await this.api!.rpc.chain.getBlock(blockHash);

        testLog.getLog().info(signedBlock.block.header.hash.toHex());

        // the hash for each extrinsic in the block
        signedBlock.block.extrinsics.forEach((ex, index) => {
          testLog.getLog().info(index + JSON.stringify(ex.toHuman()));
          this.systemExtrinics.push(ex);
        });
      },
    );
  }
  async stop(): Promise<void> {
    this.subscribeToHead();
  }

  public prettyPrint(): string {
    return `
    ________________________________________
    | ${this.name}                          
    | ${this.wsPath}                        
    |_______________________________________
    | Block connected | ${this.firstBlock}  
    | Latest block    | ${this.lastBlock}   
    | Latest hash     | ${this.lastHash}    
    |_______________________________________`;
  }
}
