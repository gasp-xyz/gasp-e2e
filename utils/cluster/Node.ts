import { ApiPromise } from "@polkadot/api";
import { initApi } from "../api";
import { testLog } from "../Logger";

class Node {
  name: string;
  connected: boolean;
  wsPath: string;
  api?: ApiPromise;

  firstBlock?: number;
  lastBlock?: number;
  lastHash?: string;
  hashes: Set<string> = new Set();
  blockHashes: Map<number, string> = new Map();
  blockNumbers: Set<number> = new Set();
  subscription: any;

  constructor(name: string, wsPath: string) {
    this.name = name;
    this.wsPath = wsPath;
    this.connected = false;
  }

  async connect(): Promise<void> {
    if (this.connected) {
      return;
    }
    this.api = await initApi(this.wsPath);
    this.connected = true;
  }

  async subscribeToHead(): Promise<void> {
    if (!this.connected) {
      throw new Error("The node is not connected yet.");
    }
    this.subscription = await this.api!.rpc.chain.subscribeNewHeads(
      (lastHeader) => {
        if (!this.firstBlock) {
          this.firstBlock = lastHeader.number.toNumber();
        }
        this.lastBlock = lastHeader.number.toNumber();
        this.lastHash = lastHeader.hash.toString();
        this.hashes.add(this.lastHash);
        this.blockNumbers.add(this.lastBlock);
        this.blockHashes.set(this.lastBlock, this.lastHash);
        testLog
          .getLog()
          .info(`${this.name} - #${this.lastBlock} - ${this.lastHash}`);
      },
    );
  }

  async stop(): Promise<void> {
    if (!this.connected) {
      throw new Error("The node is not connected yet.");
    }
    this.subscribeToHead();
  }

  prettyPrint(): string {
    return `
    ________________________________________
    | ${this.name}                          
    | ${this.wsPath}                        
    |_______________________________________
    | Connected       | ${this.connected}   
    | Block connected | ${this.firstBlock}  
    | Latest block    | ${this.lastBlock}   
    | Latest hash     | ${this.lastHash}    
    |_______________________________________`;
  }
}
export { Node };
