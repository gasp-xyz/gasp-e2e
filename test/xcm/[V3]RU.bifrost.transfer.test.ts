import { BuildBlockMode, connectParachains } from "@acala-network/chopsticks";
import { BN_BILLION, BN_HUNDRED, Mangata } from "@mangata-finance/sdk";
import { BN_FIVE, BN_TEN, BN } from "@polkadot/util";
import { mangataChopstick } from "../../utils/api";
import {
  AssetId,
  ChainId,
  ChainSpecs,
  TRANSFER_INSTRUCTIONS,
} from "../../utils/ChainSpecs";
import { waitForEvents } from "../../utils/eventListeners";
import { XcmNode } from "../../utils/Framework/Node/XcmNode";
import { ApiContext } from "../../utils/Framework/XcmHelper";
import XcmNetworks from "../../utils/Framework/XcmNetworks";
import { alice, api, setupApi, setupUsers } from "../../utils/setup";
import { expectEvent, expectJson } from "../../utils/validators";
import { Assets } from "../../utils/Assets";

/**
 * @group xcm
 */
describe("XCM transfers", () => {
  let bifrost: ApiContext;
  let mangata: ApiContext;
  let bifrostApi: XcmNode;

  beforeAll(async () => {
    bifrost = await XcmNetworks.biforst({
      buildBlockMode: BuildBlockMode.Instant,
    });
    await setupApi();
    mangata = mangataChopstick!;
    await connectParachains([bifrost.chain, mangata.chain]);

    bifrostApi = new XcmNode(bifrost.api, ChainId.Bifrost);
    setupUsers();
  });

  beforeEach(async () => {
    await mangata.dev.setStorage({
      Sudo: {
        Key: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      },
      Tokens: {
        Accounts: [
          [
            [alice.keyRingPair.address, { token: 0 }],
            { free: BN_BILLION.mul(AssetId.Mgx.unit).toString() },
          ],
          [
            [alice.keyRingPair.address, { token: 14 }],
            { free: BN_BILLION.mul(AssetId.BncV3.unit).toString() },
          ],
          [
            [alice.keyRingPair.address, { token: 4 }],
            { free: BN_BILLION.mul(AssetId.Mgx.unit).toString() },
          ],
        ],
      },
    });
    await bifrost.dev.setStorage({
      System: {
        Account: [
          [
            [alice.keyRingPair.address],
            { data: { free: BN_HUNDRED.mul(AssetId.BncV3.unit).toString() } },
          ],
        ],
      },
      Tokens: {
        Accounts: [
          [
            [alice.keyRingPair.address, { token: "ZLK" }],
            { free: BN_BILLION.mul(AssetId.Mgx.unit).toString() },
          ],
          [
            [alice.keyRingPair.address, { VToken: "ksm" }],
            { free: BN_BILLION.mul(BN_TEN.pow(new BN(12))).toString() },
          ],
          [
            [alice.keyRingPair.address, { VToken: "BNC" }],
            { free: BN_BILLION.mul(AssetId.BncV3.unit).toString() },
          ],
          [
            [alice.keyRingPair.address, { VSToken: "ksm" }],
            { free: BN_BILLION.mul(BN_TEN.pow(new BN(12))).toString() },
          ],
        ],
      },
    });
    // await upgradeMangata(mangata);

    //TODO: Remove when clarified how to setup tokens on Bifrost.
    const mgaSdk = Mangata.instance([mangata.uri]);
    await mgaSdk.xTokens.withdraw({
      account: alice.keyRingPair,
      amount: AssetId.Bnc.unit.mul(BN_HUNDRED.add(BN_TEN)),
      destinationAddress: alice.keyRingPair.address,
      parachainId: 2001,
      tokenSymbol: "BNC",
      withWeight:
        TRANSFER_INSTRUCTIONS * ChainSpecs.get(ChainId.Bifrost)!.unitCostWeight,
    });
    await bifrost.chain.newBlock();
    await bifrost.chain.newBlock();
    //END-TODO
  });

  it("[ BNC V3 -> MGA -> BNC V3 ] send BNC to mangata and back", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    const target = ChainSpecs.get(ChainId.Mg)!;
    const asset = bifrostApi.chain.assets.get(AssetId.ImbueBncV3)!;
    await mgaSdk.xTokens.depositFromParachain({
      account: alice.keyRingPair,
      asset: {
        V3: {
          id: {
            Concrete: asset.location,
          },
          fun: {
            Fungible: AssetId.Bnc.unit.mul(BN_HUNDRED),
          },
        },
      },
      destination: {
        V3: {
          parents: 1,
          interior: {
            X2: [
              { Parachain: target.parachain },
              {
                AccountId32: {
                  network: undefined,
                  id: alice.keyRingPair.addressRaw,
                },
              },
            ],
          },
        },
      },
      url: bifrost.uri,
      weightLimit: {
        Limited: {
          refTime: TRANSFER_INSTRUCTIONS * target.unitCostWeight,
          proofSize: 0,
        },
      },
    });

    //     const op = bifrostApi.xTokenTransferV3(
    //       ChainId.Mg,
    //       AssetId.ImbueBncV3,
    //       AssetId.ImbueBncV3.unit.mul(BN_TEN),
    //       alice
    //     );
    //     await signSendSuccess(bifrost.api, op, alice);

    await waitForEvents(api, "xcmpQueue.Success");

    expectEvent(await waitForEvents(api, "tokens.Deposited"), {
      event: expect.objectContaining({
        data: expect.objectContaining({
          who: "5EYCAe5ijiYfyeZ2JJCGq56LmPyNRAKzpG4QkoQkkQNB5e6Z",
          amount: "25,804,800,000",
        }),
      }),
    });
    await mgaSdk.xTokens.withdraw({
      account: alice.keyRingPair,
      amount: AssetId.Bnc.unit.mul(BN_FIVE),
      destinationAddress: alice.keyRingPair.address,
      parachainId: 2001,
      tokenSymbol: "BNC",
      withWeight:
        TRANSFER_INSTRUCTIONS * ChainSpecs.get(ChainId.Bifrost)!.unitCostWeight,
    });
    /**
    await api.tx.xTokens
      .transferMultiasset(
        {
          V3: {
            id: {
              Concrete: AssetId.BncV3.location,
            },
            fun: {
              Fungible: AssetId.Bnc.unit.mul(BN_FIVE),
            },
          },
        },
        {
          V3: {
            parents: 1,
            interior: {
              X2: [
                { Parachain: ChainSpecs.get(ChainId.Bifrost)!.parachain },
                {
                  AccountId32: {
                    network: undefined,
                    id: alice.keyRingPair.publicKey,
                  },
                },
              ],
            },
          },
        },
        {
          Limited: {
            refTime:
              TRANSFER_INSTRUCTIONS *
              ChainSpecs.get(ChainId.Bifrost)!.unitCostWeight,
            proofSize: 0,
          },
        }
      )
      .signAndSend(alice.keyRingPair);
  */
    await waitForEvents(api, "system.ExtrinsicSuccess");
    await waitForEvents(bifrost.api, "xcmpQueue.Success");

    expectEvent(await waitForEvents(bifrost.api, "balances.Deposit"), {
      event: expect.objectContaining({
        data: expect.objectContaining({
          who: "eCSrvbA5gGNYdM3UjBNxcBNBqGxtz3SEEfydKragtL4pJ4F",
          amount: "5,631,360,000",
        }),
      }),
    });
    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 14),
    ).toMatchSnapshot();
  });
  it("[ BNC V3 -> MGA -> BNC V3 ] send ZLK to mangata and back", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    const target = ChainSpecs.get(ChainId.Mg)!;
    await mgaSdk.xTokens.depositFromParachain({
      account: alice.keyRingPair,
      asset: {
        V3: {
          id: {
            Concrete: {
              parents: 0,
              interior: {
                X1: {
                  GeneralKey: {
                    length: 2,
                    data: "0x0207000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              },
            },
          },
          fun: {
            Fungible: BN_TEN.mul(Assets.MG_UNIT),
          },
        },
      },
      destination: {
        V3: {
          parents: 1,
          interior: {
            X2: [
              { Parachain: target.parachain },
              {
                AccountId32: {
                  network: undefined,
                  id: alice.keyRingPair.addressRaw,
                },
              },
            ],
          },
        },
      },
      url: bifrost.uri,
      weightLimit: {
        Limited: {
          refTime: TRANSFER_INSTRUCTIONS * target.unitCostWeight,
          proofSize: 0,
        },
      },
    });
    await waitForEvents(api, "xcmpQueue.Success");
    expectEvent(await waitForEvents(api, "tokens.Deposited"), {
      event: expect.objectContaining({
        data: expect.objectContaining({
          who: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          amount: "9,951,616,000,000,000,000",
        }),
      }),
    });
    await mgaSdk.xTokens.withdraw({
      account: alice.keyRingPair,
      amount: Assets.MG_UNIT.mul(BN_FIVE),
      destinationAddress: alice.keyRingPair.address,
      parachainId: 2001,
      tokenSymbol: "ZLK",
      withWeight:
        TRANSFER_INSTRUCTIONS * ChainSpecs.get(ChainId.Bifrost)!.unitCostWeight,
    });
    await waitForEvents(api, "system.ExtrinsicSuccess");
    await waitForEvents(bifrost.api, "xcmpQueue.Success");
    expectEvent(await waitForEvents(bifrost.api, "tokens.Deposited"), {
      event: expect.objectContaining({
        data: expect.objectContaining({
          who: "gXCcrjjFX3RPyhHYgwZDmw8oe4JFpd5anko3nTY8VrmnJpe",
          amount: "4,989,441,200,000,000,000",
          currencyId: expect.objectContaining({
            Token: "ZLK",
          }),
        }),
      }),
    });
    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 26),
    ).toMatchSnapshot();
  });
  it("[ BNC V3 -> MGA -> BNC V3 ] send vKSM to mangata and back", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    const target = ChainSpecs.get(ChainId.Mg)!;
    await mgaSdk.xTokens.depositFromParachain({
      account: alice.keyRingPair,
      asset: {
        V3: {
          id: {
            Concrete: {
              parents: 0,
              interior: {
                X1: {
                  GeneralKey: {
                    length: 2,
                    data: "0x0104000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              },
            },
          },
          fun: {
            Fungible: BN_TEN.mul(BN_TEN.pow(new BN(12))),
          },
        },
      },
      destination: {
        V3: {
          parents: 1,
          interior: {
            X2: [
              { Parachain: target.parachain },
              {
                AccountId32: {
                  network: undefined,
                  id: alice.keyRingPair.addressRaw,
                },
              },
            ],
          },
        },
      },
      url: bifrost.uri,
      weightLimit: {
        Limited: {
          refTime: 800000000,
          proofSize: 0,
        },
      },
    });
    await waitForEvents(api, "xcmpQueue.Success");
    expectEvent(await waitForEvents(api, "tokens.Deposited"), {
      event: expect.objectContaining({
        data: expect.objectContaining({
          who: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          amount: "9,999,677,440,000",
        }),
      }),
    });
    await mgaSdk.xTokens.withdraw({
      account: alice.keyRingPair,
      amount: BN_FIVE.mul(BN_TEN.pow(new BN(12))),
      destinationAddress: alice.keyRingPair.address,
      parachainId: 2001,
      tokenSymbol: "vKSM",
      withWeight:
        TRANSFER_INSTRUCTIONS * ChainSpecs.get(ChainId.Bifrost)!.unitCostWeight,
    });
    await waitForEvents(api, "system.ExtrinsicSuccess");
    await waitForEvents(bifrost.api, "xcmpQueue.Success");
    expectEvent(await waitForEvents(bifrost.api, "tokens.Deposited"), {
      event: expect.objectContaining({
        data: expect.objectContaining({
          who: "gXCcrjjFX3RPyhHYgwZDmw8oe4JFpd5anko3nTY8VrmnJpe",
          amount: "4,999,929,608,000",
          currencyId: expect.objectContaining({
            VToken: "KSM",
          }),
        }),
      }),
    });
    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 15),
    ).toMatchSnapshot();
  });
  it("[ BNC V3 -> MGA -> BNC V3 ] send vsKSM to mangata and back", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    const target = ChainSpecs.get(ChainId.Mg)!;
    await mgaSdk.xTokens.depositFromParachain({
      account: alice.keyRingPair,
      asset: {
        V3: {
          id: {
            Concrete: {
              parents: 0,
              interior: {
                X1: {
                  GeneralKey: {
                    length: 2,
                    data: "0x0404000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              },
            },
          },
          fun: {
            Fungible: BN_TEN.mul(BN_TEN.pow(new BN(12))),
          },
        },
      },
      destination: {
        V3: {
          parents: 1,
          interior: {
            X2: [
              { Parachain: target.parachain },
              {
                AccountId32: {
                  network: undefined,
                  id: alice.keyRingPair.addressRaw,
                },
              },
            ],
          },
        },
      },
      url: bifrost.uri,
      weightLimit: {
        Limited: {
          refTime: 800000000,
          proofSize: 0,
        },
      },
    });
    await waitForEvents(api, "xcmpQueue.Success");
    expectEvent(await waitForEvents(api, "tokens.Deposited"), {
      event: expect.objectContaining({
        data: expect.objectContaining({
          who: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          amount: "9,999,677,440,000",
        }),
      }),
    });
    await mgaSdk.xTokens.withdraw({
      account: alice.keyRingPair,
      amount: BN_FIVE.mul(BN_TEN.pow(new BN(12))),
      destinationAddress: alice.keyRingPair.address,
      parachainId: 2001,
      tokenSymbol: "vsKSM",
      withWeight:
        TRANSFER_INSTRUCTIONS * ChainSpecs.get(ChainId.Bifrost)!.unitCostWeight,
    });
    await waitForEvents(api, "system.ExtrinsicSuccess");
    await waitForEvents(bifrost.api, "xcmpQueue.Success");
    expectEvent(await waitForEvents(bifrost.api, "tokens.Deposited"), {
      event: expect.objectContaining({
        data: expect.objectContaining({
          who: "gXCcrjjFX3RPyhHYgwZDmw8oe4JFpd5anko3nTY8VrmnJpe",
          amount: "4,999,929,608,000",
          currencyId: expect.objectContaining({
            VSToken: "KSM",
          }),
        }),
      }),
    });
    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 16),
    ).toMatchSnapshot();
  });
});
