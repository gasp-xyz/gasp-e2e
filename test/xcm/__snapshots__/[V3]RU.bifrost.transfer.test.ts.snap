// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XCM transfers [ BNC V3 -> MGA -> BNC V3 ] send BNC to mangata and back 1`] = `
{
  "free": "0x000000000000003635c9a01b66de1000",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM transfers [ BNC V3 -> MGA -> BNC V3 ] send ZLK to mangata and back 1`] = `
{
  "free": "0x000000000000000044b7ac8468240000",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM transfers [ BNC V3 -> MGA -> BNC V3 ] send vKSM to mangata and back 1`] = `
{
  "free": 4999677440000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM transfers [ BNC V3 -> MGA -> BNC V3 ] send vsKSM to mangata and back 1`] = `
{
  "free": 4999677440000,
  "frozen": 0,
  "reserved": 0,
}
`;
